typing/typemod.ml

```diff
@@ -1163,6 +1167,48 @@ let rec approx_modtype env smty =
       let aliasable = (not (Env.is_functor_arg path env)) in
       Mty_strengthen (mty, path, Aliasability.aliasable aliasable)
 
+and approx_with ~loc env sg constr =
+  let open struct
+    type namespace = Type | Module | Module_type
+  end in
+  let rec remove_from_sg namespace path env sg =
+    List.filter_map (remove_from_sig_item namespace path env sg) sg
+  and remove_from_sig_item namespace path env sg_for_env item =
+    match path, namespace, item with
+    | [s], Type, Sig_type (id, _, _, _)
+        when Ident.name id = s ->
+      None
+    | [s], Module, Sig_module (id, _, _, _, _)
+        when Ident.name id = s ->
+      None
+    | [s], Module_type, Sig_modtype (id, _, _)
+        when Ident.name id = s ->
+      None
+    | s :: path, namespace,
+      Sig_module (id, presence, md, rs, priv)
+        when Ident.name id = s ->
+      let env = Env.add_signature sg_for_env env in
+      let sg = extract_sig env loc md.md_type in
+      let md =
+        { md with
+          md_type = Mty_signature (remove_from_sg namespace path env sg) }
+      in
+      Some (Sig_module (id, presence, md, rs, priv))
+    | _ ->
+      Some item
+  in
+  (* Removing destructively-substituted items approximates [transl_with],
+     as all approximated types are treated as abstract. *)
```

Comment:
The obvious worry here is that we could get weird dangling references.  If you consider extending `S` in your first example such that it contains references to the module `A` that's about to be removed, do we end up getting a bad approximate signature with references to items that no longer exist, when we approximate `S with A := ...`?

I think the answer is no: the comment above `approx_modtype` says that the only things it keeps are type, module, and module type elements.  Since all the types are made abstract, as you say, there is no place in the approximated signature for a dangling reference to `A` to appear.

Can you flesh out this comment to explain in more detail why this is OK?